src/net/neighbors/db.rs

```diff
@@ -0,0 +1,652 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2023 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::net::db::PeerDB;
+
+use crate::net::{
+    Error as net_error, HandshakeAcceptData, HandshakeData, Neighbor, NeighborAddress, NeighborKey,
+    Preamble, StackerDBHandshakeData, StacksMessage,
+};
+
+use crate::net::neighbors::{NeighborWalkResult, NEIGHBOR_MINIMUM_CONTACT_INTERVAL, NUM_NEIGHBORS};
+
+use crate::net::p2p::PeerNetwork;
+
+use crate::util_lib::db::DBConn;
+use crate::util_lib::db::DBTx;
+
+use std::cmp;
+use std::mem;
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use crate::burnchains::Address;
+use crate::burnchains::Burnchain;
+use crate::burnchains::BurnchainView;
+
+use rand::prelude::*;
+use rand::thread_rng;
+use stacks_common::util::get_epoch_time_secs;
+use stacks_common::util::hash::Hash160;
+use stacks_common::util::log;
+
+/// Capture replacement state
+#[derive(Debug, Clone, PartialEq)]
+pub struct NeighborReplacements {
+    /// neighbors to be replaced
+    replacements: HashMap<NeighborAddress, Neighbor>,
+    /// slots in the peer DB into which a neighbor will be stored if it must be replaced
+    replaced_neighbors: HashMap<NeighborAddress, u32>,
+}
+
+impl NeighborReplacements {
+    pub fn new() -> NeighborReplacements {
+        NeighborReplacements {
+            replacements: HashMap::new(),
+            replaced_neighbors: HashMap::new(),
+        }
+    }
+
+    pub fn add_neighbor(&mut self, naddr: NeighborAddress, neighbor: Neighbor, slot: u32) {
+        self.replacements.insert(naddr.clone(), neighbor);
+        self.replaced_neighbors.insert(naddr, slot);
```

Comment:
What if `slot` has already been inserted here? That is, the same neighbor is slated to be replaced by two new neighbors. I don't think that leads to any incoherence in the peer state, because the second new neighbor just replaces the first new neighbor during `replace_neighbors()`, but is that the intended behavior?

Reply:
It's not intentional, but I don't think it warrants a fix.  Slot collisions are highly unlikely but not astronomically so -- per the birthday paradox, you'd need about 2^16 nodes inserted before the probability of a slot collision exceeds 50%.  Even if there's a collision, there's no loss of coherence.